File: 3Dify/3Dify/Parallax/Encoders/Blur/BlurPassEncoder.m
//
//  BlurPassEncoder.m
//  3Dify
//
//  Created by It's free real estate on 02.04.20.
//  Copyright © 2020 Philipp Matthes. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "BlurPassEncoder.h"
#import <simd/simd.h>

typedef struct {
    simd_bool isVertical;
    simd_float1 blurIntensity;
    simd_float1 focalPoint;
} BlurUniforms;

@interface BlurPassEncoder ()
@property (nonatomic, strong) id<MTLDevice> device;
@property (nonatomic, strong) id<MTLRenderPipelineState> pipelineState;
@property (nonatomic, strong) id<MTLBuffer> blurUniformsBuffer;
@property (nonatomic) simd_bool isVertical;
@property (nonatomic) simd_float1 blurIntensity;
@property (nonatomic) simd_float1 focalPoint;
@end

@implementation BlurPassEncoder

-(instancetype)initWithDevice:(id<MTLDevice>)device
                   isVertical:(simd_bool)isVertical
{
    self = [super init];
    if (self) {
        self.device = device;
        self.pipelineState = [self blurPipelineStateOnDevice:device];
        self.blurUniformsBuffer = [self makeBlurUniformsBuffer];
        self.isVertical = isVertical;
        self.blurIntensity = 5.0f;
        self.focalPoint = 0.0f;
    }
    return self;
}

-(id<MTLBuffer>)makeBlurUniformsBuffer
{
    id<MTLBuffer> buffer = [self.device newBufferWithLength:sizeof(BlurUniforms)
                                                    options:MTLResourceOptionCPUCacheModeDefault];
    buffer.label = @"Blur Uniforms";
    return buffer;
}

-(id<MTLRenderPipelineState>)blurPipelineStateOnDevice:(id<MTLDevice>)device
{
    id<MTLLibrary> library = [device newDefaultLibrary];
    MTLRenderPipelineDescriptor *descriptor = [MTLRenderPipelineDescriptor new];
    descriptor.label = @"Blur Pipeline State";
    descriptor.vertexFunction = [library newFunctionWithName:@"project_texture"];
    descriptor.fragmentFunction = [library newFunctionWithName:@"blur"];
    descriptor.colorAttachments[0].pixelFormat = MTLPixelFormatRGBA16Float;
    descriptor.colorAttachments[1].pixelFormat = MTLPixelFormatRGBA16Float;
    descriptor.depthAttachmentPixelFormat = MTLPixelFormatInvalid;
    NSError *error = nil;
    id<MTLRenderPipelineState> pipelineState = [device newRenderPipelineStateWithDescriptor:descriptor error:&error];
    if (!pipelineState) {
        NSLog(@"Error occurred when creating render pipeline state: %@", error);
    }
    return pipelineState;
}

-(void)  encodeIn:(id<MTLCommandBuffer>)commandBuffer
inputColorTexture:(id<MTLTexture>)inputColorTexture
inputDepthTexture:(id<MTLTexture>)inputDepthTexture
   blurredTexture:(id<MTLTexture>)blurredTexture
     drawableSize:(CGSize)drawableSize
       clearColor:(MTLClearColor)clearColor
{
    MTLRenderPassDescriptor* descriptor = [
   self outputToColorTextureDescriptorOfSize:drawableSize
                                  clearColor:clearColor
                                  blurredTexture:blurredTexture];
    [self updateBlurUniforms];
    id<MTLRenderCommandEncoder> encoder = [commandBuffer renderCommandEncoderWithDescriptor:descriptor];
    [encoder setLabel:@"Blur"];
    [encoder setRenderPipelineState:self.pipelineState];
    [encoder setFragmentTexture:inputColorTexture atIndex:0];
    [encoder setFragmentTexture:inputDepthTexture atIndex:1];
    [encoder setFragmentBuffer:self.blurUniformsBuffer offset:0 atIndex:0];
    [encoder drawPrimitives:MTLPrimitiveTypeTriangleStrip vertexStart:0 vertexCount:4];
    [encoder endEncoding];
}

-(void)updateBlurIntensity:(simd_float1)blurIntensity
{
    self.blurIntensity = blurIntensity;
}

-(void)updateFocalPoint:(simd_float1)focalPoint
{
    self.focalPoint = focalPoint;
}

-(void)updateBlurUniforms
{
    BlurUniforms uniforms = (BlurUniforms) {
        .isVertical = self.isVertical,
        .blurIntensity = self.blurIntensity,
        .focalPoint = self.focalPoint
    };
    memcpy(self.blurUniformsBuffer.contents, &uniforms, sizeof(BlurUniforms));
}

-(MTLRenderPassDescriptor *)outputToColorTextureDescriptorOfSize:(CGSize)size
                                                      clearColor:(MTLClearColor)clearColor
                                                  blurredTexture:(id<MTLTexture>)blurredTexture
{
    MTLRenderPassDescriptor *descriptor = [MTLRenderPassDescriptor renderPassDescriptor];
    descriptor.colorAttachments[0].texture = blurredTexture;
    descriptor.colorAttachments[0].clearColor = clearColor;
    descriptor.colorAttachments[0].loadAction = MTLLoadActionLoad;
    descriptor.colorAttachments[0].storeAction = MTLStoreActionStore;
    descriptor.renderTargetWidth = size.width;
    descriptor.renderTargetHeight = size.height;
    return descriptor;
}

@end

File: ThreeDify/ThreeDify/Parallax/Encoders/BlurPassEncoder.swift
//
// 3Dify App
//
// Project website: https://github.com/3dify-app
//
// Authors:
// - Philipp Matthes 2020, Contact: mail@philippmatth.es
//
// Copyright notice: All rights reserved by the authors given above. Do not
// remove or change this copyright notice without confirmation of the authors.
//

import Foundation
import MetalKit
import simd

struct BlurPassUniforms {
    var isVertical: simd_bool
    var blurIntensity: simd_float1
    var focalPoint: simd_float1
    
    init(
        isVertical: simd_bool,
        blurIntensity: simd_float1 = 5,
        focalPoint: simd_float1 = 0
    ) {
        self.isVertical = isVertical
        self.blurIntensity = blurIntensity
        self.focalPoint = focalPoint
    }
}

struct BlurPassEncoder {
    private let device: MTLDevice
    private let pipelineState: MTLRenderPipelineState
    private let uniformsBuffer: MTLBuffer
    
    var uniforms: BlurPassUniforms
    
    enum InitializationError: Error {
        case libraryInitFailed
        case pipelineInitFailed
        case bufferInitFailed
    }
    
    enum RenderingError: Error {
        case makeDescriptorFailed
    }
    
    init(device: MTLDevice, isVertical: Bool) throws {
        try self.init(device: device, uniforms: .init(isVertical: isVertical))
    }
    
    init(device: MTLDevice, uniforms: BlurPassUniforms) throws {
        self.device = device
        
        guard
            let library = device.makeDefaultLibrary()
        else { throw InitializationError.libraryInitFailed }
        
        let descriptor = MTLRenderPipelineDescriptor()
        descriptor.label = "\(uniforms.isVertical ? "Vertical" : "Horizontal") Blur Pipeline State"
        descriptor.vertexFunction = library.makeFunction(name: "project_texture")
        descriptor.fragmentFunction = library.makeFunction(name: "blur")
        descriptor.colorAttachments[0].pixelFormat = .rgba16Float
        descriptor.colorAttachments[1].pixelFormat = .rgba16Float
        descriptor.depthAttachmentPixelFormat = .invalid
        
        guard
            let pipelineState = try? device.makeRenderPipelineState(
                descriptor: descriptor
            )
        else { throw InitializationError.pipelineInitFailed }
        self.pipelineState = pipelineState
        guard
            let buffer = device.makeBuffer(
                length: MemoryLayout.size(ofValue: uniforms),
                options: .cpuCacheModeWriteCombined
            )
        else { throw InitializationError.bufferInitFailed }
        buffer.label = "\(uniforms.isVertical ? "Vertical" : "Horizontal") Blur Uniforms Buffer"
        self.uniformsBuffer = buffer
        
        self.uniforms = uniforms
    }
    
    mutating func encode(
        inCommandBuffer commandBuffer: MTLCommandBuffer,
        inputColorTexture: MTLTexture,
        inputDepthTexture: MTLTexture,
        outputBlurredTexture: MTLTexture,
        drawableSize: CGSize,
        clearColor: MTLClearColor
    ) throws {
        let descriptor = MTLRenderPassDescriptor()
        descriptor.colorAttachments[0].texture = outputBlurredTexture
        descriptor.colorAttachments[0].clearColor = clearColor
        descriptor.colorAttachments[0].loadAction = .load
        descriptor.colorAttachments[0].storeAction = .store
        descriptor.renderTargetWidth = Int(drawableSize.width)
        descriptor.renderTargetHeight = Int(drawableSize.height)
        memcpy(
            uniformsBuffer.contents(),
            &uniforms,
            MemoryLayout.size(ofValue: uniforms)
        )
        guard
            let encoder = commandBuffer.makeRenderCommandEncoder(
                descriptor: descriptor
            )
        else { throw RenderingError.makeDescriptorFailed }
        encoder.label = "\(uniforms.isVertical ? "Vertical" : "Horizontal") Blur"
        encoder.setRenderPipelineState(pipelineState)
        encoder.setFragmentTexture(inputColorTexture, index: 0)
        encoder.setFragmentTexture(inputDepthTexture, index: 1)
        encoder.setFragmentBuffer(uniformsBuffer, offset: 0, index: 0)
        encoder.drawPrimitives(type: .triangleStrip, vertexStart: 0, vertexCount: 4)
        encoder.endEncoding()
    }
}

